proposals/devcontainer-features.md

```diff
@@ -0,0 +1,143 @@
+
+In addition to the list of features included, the file includes the following properties.
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| name | string | Name of the collection |
+| reference | string | Reference information of the repository and path where the code is stored. |
+| version | string | Version of the code. |
+
+In most cases the `feature-collection.json` file can be generated automatically at the moment of creating a release.
+
+## devcontainer.json properties
+
+Features are referenced in `devcontainer.json`  where the `features` tag consists of an array with the ordered list of features to be included in the container image.
+
+The properties are:
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| id | string | reference to the particular feature to be included. |
+| options | object | Type of the option |
+
+The `id` is the main reference point for how to find and download a particular feature. `id` can be defined in any of the following ways:
+
+| Type | Description |
+| :--- | :--- |
+| feature  | A simple name references a feature included with the application used to execute `devcontainer.json` |
+| organization/repository/{feature or collectionl/feature}/@version | Reference to a particular release in a repository |
+| https://<../URI/..>/devcontainer-features.tgz#{optional feature} | Direct reference to a file download. |
+| ./{local-path}  -or-  ../{local-path} | A path relative to devcontainer.json where a feature or feature collection can be found. |
+
+## Authoring
+
+Features can be authored in a number of languages, the most straight forward being bash scripts. If a feature is authored in a different language information about it should be included in the metadata so that users can make an informed choice about it.
+
+Reference information about the application required to execute the feature should be included in `feature.json` in the metadata section.
+
+Applications should default to `/bin/sh` for features that do not include this information.
+
+If the feature is included in a folder as part of the repository that contains `devcontainer.json` no other steps are necessary.
+
+## Release
+
+A release is created when the objective is to have other users use a feature.
+
+A release consists of the following:
+
+1.- Tar file with all the included files for the feature or feature collection.
+2.- A copy of the `feature.json` or `feature-collection.json` file that defines the contents of the tar file with additional information added to validate it:
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| version | string | SemVer version of the release. |
+| checksum | string | Checksum of the tar file. |
+
+## Execution
+
+The application that implements features should:
+
+- Features are executed in the order defined in devcontainer.json
```

Comment:
We should keep this declarative for several reasons:
- Usability: When there are implicit dependencies the user has no way knowing what they are and we have no way of changing them over time without risking to break existing configurations.
- Parallelization: We want to preserve the option to parallelize installation in the future, using multi-stage builds or otherwise.
- Explicit dependencies: We want to preserve the option to add explicit dependencies, these would change the order of installation (potentially change it over time as dependencies change).

One example was the creation of a user: Since the user is specified in the devcontainer.json as a top-level option, it makes more sense to deal with that in general instead of as part of a feature (the `common` feature?). It might be easiest if we create the user before installing any features if that user does not exist yet.

Reply:
- I like the idea of having/using the cached features for things like users, although I'm not sure if defaulting to running it before other features is the best idea.
I feel like there are two ways of dealing with this. One is the array format which makes it clear to the user in what order things are going to run and the user has total control of that. Not sure I like the idea of things changing from under me if there is an update somewhere upstream. Declaring dependencies could actually be done in `devcontainer.json` instead of as part of the feature itself. If we do that the tool can more easily parallelize, but it complicates the user experience in that now they have to declare the dependencies instead of just using an in order execution.

Thoughts? @Chuxel @jkeech @joshspicer

Reply:
To clarify: Dependencies would be declared by features themselves (not in the devcontainer.json), the user should not have to know about these. We don't have dependencies at the moment, but since we have already investigated adding them, I use them as one example to show that having `"features"` in the devcontainer.json determine an execution order will constrain us in the future. (Even more important is the usability argument above.)

On the user example: The user is configured with the top-level `"remoteUser"` independently of features. That suggests that we should address it also independently of features. Since features then depend on that user, we would create it before we run the feature install scripts. (Having a feature create the `"remoteUser"` doesn't seem to make sense since that user needs to exist also without any features for the configuration to be valid.)

Another argument for keeping "features" declarative is that it will make adding auto-detection of features in the future simpler. (If order matters, we need logic to arrange that for the detected features.)

Reply:
The main idea of making this changes is to allow users to control the order of execution, especially since we are not supporting dependencies yet and the changes to the OS applied by features might clash between themselves in unpredictable ways that a user might need to fix in this way. 

Thoughts @Chuxel, @bamurtaugh, @jkeech, @joshspicer ?

Reply:
I'm not sure if this would work, but is there a way a user could set a property if the order of their features matters? I think the concerns @chrmarti mentions make sense, but I can see the value in what @edgonmsft mentions about controlling the order of execution, so could we default to declarative but allow setting a property that ordering matters, in case users would like to override this declarative default?

Reply:
The order that works best will depend on how the features are implemented and could change over time. The user does not have any insight into this. When we add dependencies we want to be in control of the order.

We should ensure the order of installation is always the same (e.g., sort by id), so the outcome is reproducible. If we have any implicit dependencies today, I suggest we take a closer look and find a way to deal with them short term.

Reply:
My biggest concern with obscuring the installation order of features (which are implemented as  shell scripts with no standardization on _how_ to author them), is that is will significantly hinder the ability of a community member to author their own set of features, and prevent the wider adoption and ecosystem of features. 

We have not landed on a simple, complete mechanism for automated versioning/dependency management despite a fair bit of time and effort researching and discussing options.      We've investigated building on top of existing package managers like `npm`. Our biggest detractor for such an idea is that we aren't just installing the _feature_, but the _feature with the provided options_.  Our "blessed" set of features today already function quite differently given provided options - We would need _strict_ authoring rules for the community, as well as a strictly defined set of metadata and listing of dependencies, to make the declarative features object feasible.  We are talking about reimplementing a package manager like 'apt'.

I think that is too complicated given the problem 'features' is trying to solve (An easy-to-use mechanism to get tool X available in your dev environment, that can auto-update on rebuilds if you want it to)

To the three arguments for implicit ordering:

> Usability: When there are implicit dependencies the user has no way knowing what they are and we have no way of changing them over time without risking to break existing configurations.

I think the opposite is true.  We provide the user with total control over dependencies (we hope there are few with features, but maybe there are 1 or 2).  I think _hiding_ the dependencies will cause more breaks for users.

> Parallelization: We want to preserve the option to parallelize installation in the future, using multi-stage builds or otherwise.

This argument makes sense, we lose a bit of control over installation order if we aren't able to decide that ourselves. 

> Explicit dependencies: We want to preserve the option to add explicit dependencies, these would change the order of installation (potentially change it over time as dependencies change).

This is what we are seeking with the array syntax, an option for users to add explicit dependencies.  If a dependency changes significantly, it is on the author to create a new major version of the feature and for users to make the conscious decision on how they would like to proceed in their dev containers.  I have already implemented a way to pin to a GitHub release or tgz URL, ranges are a next logical step.

Reply:
Additionally, we'll also need to take a dependency on the base image, which could be based on a yet another base image, both installing tools.  While I think, in a perfect world, we could write tooling to figure everything out for a user, pragmatically I don't think we want to get into that business :D